# Part 1. Equation of a Slime

How many late days are you using for this assignment? 0

In [281]:
# Imports section
import pandas as pd
import sklearn as sk
from numpy import ravel
from sklearn.model_selection import train_test_split, cross_val_score, ShuffleSplit
from sklearn.linear_model import LinearRegression
from sklearn.linear_model import LogisticRegression
from sklearn.preprocessing import PolynomialFeatures
from sklearn import svm
from sklearn.neighbors import KNeighborsClassifier
from sklearn.neural_network import MLPClassifier


## 1. Loading the dataset

In [282]:
# Using pandas load the dataset
# Output the first 15 rows of the data
# Display a summary of the table information (number of datapoints, etc.)

slime_data = pd.read_csv("science_data_large.csv")

slime_data.iloc[:15]

slime_data.describe()

,Temperature °C,Mols KCL,Size nm^3
count,1000.000000,1000.000000,1.000000e+03
mean,500.500000,471.530000,5.086111e+05
std,288.819436,288.482872,4.474838e+05
min,1.000000,1.000000,1.611429e+01
25%,250.750000,226.750000,1.298267e+05
50%,500.500000,459.500000,3.827182e+05
75%,750.250000,710.250000,7.603211e+05
max,1000.000000,1000.000000,1.972127e+06


## 2. Splitting the dataset

In [283]:
# Take the pandas dataset and split it into our features (X) and label (y)

labels = slime_data[["Size nm^3"]]

features = slime_data.drop(columns=["Size nm^3"])

# Use sklearn to split the features and labels into a training/test set. (90% train, 10% test)
# For grading consistency use random_state=42

features_train, features_test, labels_train, labels_test = train_test_split(
    features, labels, test_size=0.1, random_state=42)

## 3. Perform a Linear Regression

In [284]:
# Use sklearn to train a model on the training set
model1 = LinearRegression().fit(features_train, labels_train)

# Create a sample datapoint and predict the output of that sample with the trained model
data_point = [[178, 372]]
print(model1.predict(data_point))

# Report the score for that model using the default score function property of the SKLearn model, in your own words (markdown, not code) explain what the score means
print("\nScore:",model1.score(features_test, labels_test))

# Extract the coefficients and intercept from the model and write an equation for your h(x) using LaTeX
x0 = model1.coef_[0][0]
x1 = model1.coef_[0][1]
b = model1.intercept_

[[128945.14672951]]

Score: 0.8552472077276096


c:\ProgramData\anaconda3\Lib\site-packages\sklearn\base.py:493: UserWarning: X does not have valid feature names, but LinearRegression was fitted with feature names
  warnings.warn(


Write the linear equation of a slime: (example equation: $E = mc^2$)
<br>
$y = 886.14641 \cdot x0 + 1032.69507 \cdot x1 - 409391.47958$

Report on score and explain meaning:
<br>
The $R^2$ score is essentially a measurement of how well our model fits the data, a perfect fit is where every data point is in line with model predictions and has a score of 1. In this case our score ~0.85 which indicates a strong corrolation between model and data, a perfect fit may just mean overfitting which isn't good either.

## 4. Use Cross Validation

In [285]:
# Use the cross_val_score function to repeat your experiment across many shuffles of the data
# For grading consistency use n_splits=5 and random_state=42
cross_val_score(model1, features, labels, cv = ShuffleSplit(n_splits=5, test_size=0.1, random_state=42))

# Report on their finding and their significance

array([0.85524721, 0.85830729, 0.8654508 , 0.83855498, 0.83002469])

Write findings here:
<br>
The different splits did not see large improvements to the accuracy, this suggests our original model did a good job of generalizing the data and did not overfit to a particular subset of training data.

## 5. Using Polynomial Regression

In [286]:
# Using the PolynomialFeatures library perform another regression on an augmented dataset of degree 2
# Perform k-fold cross validation (as above)
poly = PolynomialFeatures(2)
p_data = poly.fit_transform(features_train, labels_train)

#print(poly.get_feature_names_out())

model2 = LinearRegression().fit(p_data, labels_train)

print(cross_val_score(model2, features, labels, cv = ShuffleSplit(n_splits=5, test_size=0.1, random_state=42)))

# Report on the metrics and output the resultant equation as you did in Part 3.
#print(model2.coef_)
#print(model2.intercept_)

[0.85524721 0.85830729 0.8654508  0.83855498 0.83002469]


Write the polynomial equation of a slime: (example equation: $E = mc^2$)
<br>
$y = 1 \cdot 0 + 1.2 \cdot x0 - 1.23111 \cdot x1 - 1.05681 \cdot x0^2 + 2 \cdot x0x1 + 2.85714 \cdot x1^2 + 1.65726$

Report on the score and interpret:
<br>
The $R^2$ scores here are the same as before.

# Part 2. Chronic Kidney Disease Prediction via Classification

Create code and markdown cells as needed to perform classification and report on your results

In [287]:
# Load the dataset. Then train and evaluate the classification models.
ckd_data = pd.read_csv("ckd_feature_subset.csv")

ckd_labels = ravel(ckd_data[["Target_ckd"]])
ckd_features = ckd_data.drop(columns=["Target_ckd"])

ckd_features_train, ckd_features_test, ckd_labels_train, ckd_labels_test = train_test_split(
    ckd_features, ckd_labels, test_size=0.1, random_state=42)

LR_model = LogisticRegression(random_state=42).fit(ckd_features_train, ckd_labels_train)
SVC_model = svm.SVC(random_state=42).fit(ckd_features, ckd_labels)
KNN_model = KNeighborsClassifier().fit(ckd_features, ckd_labels)

print(cross_val_score(LR_model, ckd_features, ckd_labels, cv = ShuffleSplit(n_splits=5, test_size=0.1, random_state=42)))
print(cross_val_score(SVC_model, ckd_features, ckd_labels, cv = ShuffleSplit(n_splits=5, test_size=0.1, random_state=42)))
print(cross_val_score(KNN_model, ckd_features, ckd_labels, cv = ShuffleSplit(n_splits=5, test_size=0.1, random_state=42)))

NN_model = MLPClassifier(random_state=42).fit(ckd_features, ckd_labels)
print(cross_val_score(NN_model, ckd_features, ckd_labels, cv = ShuffleSplit(n_splits=5, test_size=0.1, random_state=42)))


[0.6875 0.8125 0.9375 1.     0.9375]
[0.875  1.     1.     1.     0.9375]
[0.9375 1.     1.     1.     0.9375]


c:\ProgramData\anaconda3\Lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:690: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(
c:\ProgramData\anaconda3\Lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:690: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(
c:\ProgramData\anaconda3\Lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:690: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(
c:\ProgramData\anaconda3\Lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:690: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(
c:\ProgramData\anaconda3\Lib\site-packages\sklearn\neural_network\_multi

[0.75   0.9375 0.9375 1.     0.9375]


c:\ProgramData\anaconda3\Lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:690: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(


## Results and Conclusion for Classification Experiments

| Model | R² Scores                           | Average | Standard Deviation |
|-------|-------------------------------------|---------|--------------------|
| LR    | 0.6875, 0.8125, 0.9375, 1.0, 0.9375 | 0.895   | 0.1173             |
| SVC   | 0.875, 1.0, 1.0, 1.0, 0.9375        | 0.9625  | 0.0459             |
| KNN   | 0.9375, 1.0, 1.0, 1.0, 0.9375       | 0.975   | 0.025              |
| NN    | 0.75, 0.9375, 0.9375, 1.0, 0.9375   | 0.9125  | 0.0264             |

The model with the best average $R^2$ score is the k-nearest neighbors model which also has the lowest deviation. By these metrics it is the best model but I am wary of k-nearest neighbor "finding trends" that do not exist.

In [292]:
# Experiments with Neural Network.
NN_model1 = MLPClassifier(random_state=42, max_iter=1000).fit(ckd_features, ckd_labels)

NN_model2 = MLPClassifier(random_state=42, solver="sgd").fit(ckd_features, ckd_labels)

NN_model3 = MLPClassifier(random_state=42, learning_rate_init=0.005).fit(ckd_features, ckd_labels)

print(cross_val_score(NN_model1, ckd_features, ckd_labels, cv = ShuffleSplit(n_splits=5, test_size=0.1, random_state=42)))
print(cross_val_score(NN_model2, ckd_features, ckd_labels, cv = ShuffleSplit(n_splits=5, test_size=0.1, random_state=42)))
print(cross_val_score(NN_model3, ckd_features, ckd_labels, cv = ShuffleSplit(n_splits=5, test_size=0.1, random_state=42)))


c:\ProgramData\anaconda3\Lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:690: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(
c:\ProgramData\anaconda3\Lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:690: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(


[0.875  1.     1.     1.     0.9375]


c:\ProgramData\anaconda3\Lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:690: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(
c:\ProgramData\anaconda3\Lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:690: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(
c:\ProgramData\anaconda3\Lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:690: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(
c:\ProgramData\anaconda3\Lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:690: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(
c:\ProgramData\anaconda3\Lib\site-packages\sklearn\neural_network\_multi

[0.625  0.6875 0.75   0.9375 0.8125]


c:\ProgramData\anaconda3\Lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:690: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(
c:\ProgramData\anaconda3\Lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:690: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(
c:\ProgramData\anaconda3\Lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:690: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(


[0.875 1.    1.    1.    1.   ]


c:\ProgramData\anaconda3\Lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:690: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(


## Results and Conclusion for Neural Network Experiments

|  Model  | Solver | Max Iterations |  Learning Rate | R² Scores                          |
|---------|--------|----------------|----------------|------------------------------------|
| Model 1 |  adam  | 1000           | 0.001          | 0.875, 1.0, 1.0, 1.0, 0.9375       |
| Model 2 |  sgd   | 200            | 0.001          | 0.625, 0.6875, 0.75, 0.9375, 0.8125|
| Model 3 |  adam  | 200            | 0.005          | 0.875, 1.0, 1.0, 1.0, 1.0          |

For my experimentation I changed the solver, max iterations and learning rate for the model. I noticed that I was getting warnings that the model was not able to converge by the end of the default iterations of 200 so I started by increasing the max iterations to 1000. This helped but the biggest difference came from increasing the learning rate although this is not an ideal solution to improve models. Changing the solver to a standard sgd was a substantial downgrade to adam.